tiledb/sm/query/reader.cc

```diff
@@ -1054,97 +1054,159 @@ Status Reader::copy_var_cells(
   auto buffer_var = (unsigned char*)it->second.buffer_var_;
   auto buffer_size = it->second.buffer_size_;
   auto buffer_var_size = it->second.buffer_var_size_;
-  uint64_t buffer_offset = 0;
-  uint64_t buffer_var_offset = 0;
   uint64_t offset_size = constants::cell_var_offset_size;
-  uint64_t cell_var_size;
   auto type = array_schema_->type(attribute);
   auto fill_size = datatype_size(type);
   auto fill_value = constants::fill_value(type);
   assert(fill_value != nullptr);
 
-  // Copy cells
-  for (const auto& cr : cell_ranges) {
-    auto cell_num_in_range = cr.end_ - cr.start_ + 1;
-    // Check if offset buffers can fit the result
-    if (buffer_offset + cell_num_in_range * offset_size > *buffer_size) {
-      read_state_.overflowed_ = true;
-      return Status::Ok();
-    }
+  // Compute the destinations of offsets and var-len data in the buffers.
+  std::vector<std::vector<uint64_t>> offset_offsets_per_cr;
+  std::vector<std::vector<uint64_t>> var_offsets_per_cr;
+  uint64_t total_offset_size, total_var_size;
+  RETURN_NOT_OK(compute_var_cell_destinations(
+      attribute,
+      cell_ranges,
+      &offset_offsets_per_cr,
+      &var_offsets_per_cr,
+      &total_offset_size,
+      &total_var_size));
+
+  // Check for overflow and return early (without copying) in that case.
+  if (total_offset_size > *buffer_size || total_var_size > *buffer_var_size) {
+    read_state_.overflowed_ = true;
+    return Status::Ok();
+  }
 
-    // Handle empty range
-    if (cr.tile_ == nullptr) {
-      // Check if result can fit in the buffer
-      if (buffer_var_offset + cell_num_in_range * fill_size >
-          *buffer_var_size) {
-        read_state_.overflowed_ = true;
-        return Status::Ok();
-      }
+  // Copy cell ranges in parallel.
+  const auto num_cr = cell_ranges.size();
+  auto statuses = parallel_for(0, num_cr, [&](uint64_t cr_idx) {
```

Comment:
Noting here that this parallelizes across cell ranges. If the numbers of cells in the cell ranges are skewed, this will not parallelize very well.

Reply:
Hmm, that is a good point. I'm not sure how to handle that in general, although TBB may help us a bit if it is able to perform work-stealing (not sure if that is the case here).

Reply:
On a second thought, the cell ranges will not be skewed, but rather have the same number of cells.